include/realizations/catchment/Bmi_Fortran_Adapter.hpp

```diff
@@ -0,0 +1,763 @@
+             *
+             * @param name
+             * @return
+             */
+            int GetGridRank(int grid_id) override;
+
+            /**
+             * Get grid size for a grid-id.
+             *
+             * @param name
+             * @return
+             */
+            int GetGridSize(int grid_id) override;
+
+            /**
+             * Whether the backing model has been initialized yet.
+             *
+             * @return Whether the backing model has been initialized yet.
+             */
+            bool is_model_initialized();
+
+            void SetValue(std::string name, void *src) override;
+
+            template<class T>
+            void SetValue(std::string name, std::vector<T> src) {
+                size_t item_size;
+                try {
+                    item_size = (size_t) GetVarItemsize(name);
+                }
+                catch (std::runtime_error &e) {
+                    throw std::runtime_error("Cannot set " + name + " variable of " + model_name +
+                                             "; unable to test item sizes are equal (does variable exist for model?)");
+                }
+                if (item_size != sizeof(src[0])) {
+                    throw std::runtime_error("Cannot set " + name + " variable of " + model_name +
+                                             " with values of different item size");
+                }
+                SetValue(std::move(name), static_cast<void *>(src.data()));
+            }
+
+            void SetValueAtIndices(std::string name, int *inds, int count, void *src) override;
+
+            /**
+             *
+             * @tparam T
+             * @param name
+             * @param inds Collection of indexes within the model for which this variable should have a value set.
+             * @param src Values that should be set for this variable at the corresponding index in `inds`.
+             */
+
+            template<class T>
+            void SetValueAtIndices(const std::string &name, std::vector<int> inds, std::vector<T> src) {
+                if (inds.size() != src.size()) {
+                    throw std::runtime_error("Cannot set specified indexes for " + name + " variable of " + model_name +
+                                             " when index collection is different size than collection of values.");
+                }
+                if (inds.empty()) {
+                    return;
+                }
+                size_t item_size;
+                try {
+                    item_size = (size_t) GetVarItemsize(name);
+                }
+                catch (std::runtime_error &e) {
+                    throw std::runtime_error("Cannot set specified indexes for " + name + " variable of " + model_name +
+                                             "; unable to test item sizes are equal (does variable exist for model?)");
+                }
+                if (item_size != sizeof(src[0])) {
```

Comment:
What's most important is checking that the array sizes match.  That wasn't being done, but the above mentioned change to also get the number of bytes (and then with that compare the variable array size to the vector size), should guarantee the arrays match and that individual indices contain something that should be interpreted as the right value. 

This could still be used improperly if the types don't actually agree, if they are of the same item size, but I'm not sure how much we want to worry about that.

Reply:
I agree, the important part is making sure the array sizes match.  After noticing that, I got hung up thinking about how that extends to the at indicices functions.